src/EditorFeatures/Core/Shared/Extensions/ITextSnapshotExtensions.cs

```diff
@@ -63,6 +64,17 @@ private static IEnumerable<AbstractFormattingRule> GetFormattingRules(Document d
         /// </summary>
         public static Document GetFullyLoadedDocument(
             this ITextBuffer buffer, IUIThreadOperationContext operationContext)
+        {
+            return GetFullyLoadedDocument(buffer, operationContext, _ => true);
+        }
+
+        /// <summary>
+        /// Get <see cref="Document"/> from <see cref="Text.Extensions.GetDocument(ITextSnapshot)"/>
+        /// once <see cref="IWorkspaceStatusService.WaitUntilFullyLoadedAsync(CancellationToken)"/> returns.
+        /// </summary>
+        /// <param name="shouldLoad">Whether this function should wait for the solution to fully load.</param>
+        public static Document GetFullyLoadedDocument(
+            this ITextBuffer buffer, IUIThreadOperationContext operationContext, Func<Document, bool> shouldLoad)
```

Comment:
:confused: I'm not a big fan of this API. If the checks are moved, I would prefer to see them explicitly placed in the calling code before the call to `GetFullyLoadedDocumentAsync`.

Reply:
The problem with that is that you now need to get the Doucment in two different ways.  And this leads to a confusing experience in teh code because one Document should only be used for *pre-load-checks*, and one should only be used once the loading has fully happened.  

A problem discovered in the original PR was that it was super easy to do something like:

```c#
var document = ...
//cause fully load
// use 'document'
```

This is try to avoid that problem.

Reply:
I still find that situation less confusing than the one here though... Will think on this more